### PR TITLE
Seqcol comparison api

### PR DESCRIPTION
**Pull request by waterflow80**
Completed the seqCol comparison API

**Still** need to understand the [unbalanced duplicates present](https://github.com/ga4gh/seqcol-spec/blob/master/docs/decision_record.md#same-order-specification).

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/evaseqcol/controller/seqcol/SeqColComparisonController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+public class SeqColComparisonController {
+
+    private SeqColService seqColService;
+
+    @Autowired
+    public SeqColComparisonController(SeqColService seqColService) {
+        this.seqColService = seqColService;
+    }
+
+    @GetMapping("/{digest1}/{digest2}")
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/controller/seqcol/SeqColComparisonController.java:32]: **Optional**: Add links to the corresponding parts of the sequence-collection specification in a comment

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColLevelTwoService.java [hunk 2]
```diff
@@ -89,4 +89,25 @@ private List<SeqColExtendedDataEntity> getExtendedAttributes(SeqColLevelOneEntit
...
+                    levelTwoEntity.setLengths(extendedData.getExtendedSeqColData().getObject());
+                    break;
+                case names:
+                    levelTwoEntity.setNames(extendedData.getExtendedSeqColData().getObject());
+                    break;
+                case sequences:
+                    levelTwoEntity.setSequences(extendedData.getExtendedSeqColData().getObject());
+                    break;
+                case md5DigestsOfSequences:
+                    levelTwoEntity.setMd5Sequences(extendedData.getExtendedSeqColData().getObject());
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColLevelTwoService.java:107]: **Optional**: getter and setter methods should also be renamed to use the term "md5 digests of sequences"...

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 3]
```diff
@@ -115,4 +127,263 @@ private boolean isSeqColL1Present(SeqColLevelOneEntity levelOneEntity) {
...
+        List<String> seqColASequences = seqColAEntity.getSequences();
+        List<String> seqColBSequences = seqColBEntity.getSequences();
+
+        // "elements" attribute | "total"
+        setTotalElementsAttribute(seqColAEntity, seqColBEntity, comparisonResult);
+
+        // "elements" attribute | "a-and-b"
+        setA_and_B_ElementsAttribute(seqColAEntity, seqColBEntity, comparisonResult);
+
+        // "elements" attribute | "a-and-b-same-order"
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:159]: Why shouldn't the logic below refactored into a separate function as well similar to the above 2 functions?

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 4]
```diff
@@ -115,4 +127,263 @@ private boolean isSeqColL1Present(SeqColLevelOneEntity levelOneEntity) {
...
+    private <T extends SeqColEntity> void setArraysAttribute(
+            T seqColAEntity, T seqColBEntity, SeqColComparisonResultEntity comparisonResult) {
+        Field[] seqColAFields = getFields(seqColAEntity);
+        Field[] seqColBFields = getFields(seqColBEntity);
+        List<Field> seqColCommonFields = getCommonFields(seqColAEntity, seqColBEntity);
+        List<String> seqColAFieldsNames = getFieldsNames(seqColAFields);
+        List<String> seqColBFieldsNames = getFieldsNames(seqColBFields);
+        List<String> commonFieldsNames = getFieldsNames(seqColCommonFields);
+        // a-only
+        comparisonResult.putIntoArrays("a-only", seqColAFieldsNames);
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:195]: I am a bit confused... shouldn't this `A minus (A intersect B)`?

Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:195]: True, this is not correct.
I'll fix it.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 5]
```diff
@@ -6,31 +6,43 @@
...
 import uk.ac.ebi.eva.evaseqcol.entities.SeqColEntity;
 import uk.ac.ebi.eva.evaseqcol.entities.SeqColLevelOneEntity;
 import uk.ac.ebi.eva.evaseqcol.entities.SeqColExtendedDataEntity;
 import uk.ac.ebi.eva.evaseqcol.entities.SeqColLevelTwoEntity;
 import uk.ac.ebi.eva.evaseqcol.exception.DuplicateSeqColException;
+import uk.ac.ebi.eva.evaseqcol.exception.SeqColNotFoundException;
 import uk.ac.ebi.eva.evaseqcol.utils.JSONExtData;
 
 import java.io.IOException;
+import java.lang.reflect.Field;
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:21]: I am not sold on using language level reflection to do attribute comparisons across sequence collection entities.. We should discuss possible alternatives like making the Sequence collection entity flexible enough to accept multiple attributes if possible...

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java (hunk 3 above)
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:159]: In the code below we assume that we already know the fields that are present in each seqCol objects.
Which might not be the case all the time, as some seqCol might contain additional attributes in the future. That's why I wanted to make it generic.

**Review by apriltuesday — APPROVED** [review]
I think this looks good, aside from the suggestions for improvement we discussed earlier today and can do in subsequent PRs.

The reflections stuff is impressive and it's good that you included it so we could discuss it, but I'd probably prefer putting that in a separate branch and keeping this PR a bit simpler - i.e. with a less flexible compare function that only accepts seqcols with all the required attributes. Then once we determine how to best make it generic, you can do that in a PR all at once.
Just my preference though (figuring out which bits should go together in a PR is easier said than done...)

**Pull request merged**